controllers/core_clusteroperator_controller.go

```diff
@@ -0,0 +1,105 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	openshiftconfigv1 "github.com/openshift/api/config/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logr "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/openshift/platform-operators/internal/clusteroperator"
+)
+
+type CoreClusterOperatorReconciler struct {
+	client.Client
+	ReleaseVersion  string
+	SystemNamespace string
+}
+
+const (
+	coreCOName = "platform-operators-core"
+)
+
+//+kubebuilder:rbac:groups=platform.openshift.io,resources=platformoperators,verbs=list
+//+kubebuilder:rbac:groups=config.openshift.io,resources=clusteroperators,verbs=get;list;watch
+//+kubebuilder:rbac:groups=config.openshift.io,resources=clusteroperators/status,verbs=update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
+func (r *CoreClusterOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContext(ctx)
+	log.Info("reconciling request", "req", req.NamespacedName)
+	defer log.Info("finished reconciling request", "req", req.NamespacedName)
+
+	coBuilder := clusteroperator.NewBuilder()
+	coWriter := clusteroperator.NewWriter(r.Client)
+
+	core := &openshiftconfigv1.ClusterOperator{}
+	if err := r.Get(ctx, req.NamespacedName, core); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		log.Info("core clusteroperator does not exist. recreating it...", "name", coreCOName)
+		core = &openshiftconfigv1.ClusterOperator{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: coreCOName,
+			},
+			Status: openshiftconfigv1.ClusterOperatorStatus{},
+		}
+		return ctrl.Result{}, r.Create(ctx, core)
+	}
+	defer func() {
+		if err := coWriter.UpdateStatus(ctx, core, coBuilder.GetStatus()); err != nil {
+			log.Error(err, "error updating CO status")
+		}
+	}()
+
+	// Set the default ClusterOperator conditions to Available=True as the POM isn't
+	// managing any operands during phase 0.
+	coBuilder.WithDegraded(openshiftconfigv1.ConditionFalse)
+	coBuilder.WithAvailable(openshiftconfigv1.ConditionTrue, fmt.Sprintf("The platform operator manager is available at %s", r.ReleaseVersion), clusteroperator.ReasonAsExpected)
+	coBuilder.WithProgressing(openshiftconfigv1.ConditionFalse, "")
```

Comment:
Is there some sort of mechanism by which we can know to set `Progressing=true`? (e.g. when the POM or when one of the core controllers managed under the POM umbrella are being upgraded)

Reply:
Are we treating rukpak as an operand for the core ClusterOperator? I'd assume we aren't as the CVO is deploying that sub-component, and talking about this implementation in https://coreos.slack.com/archives/C030W4V9BHN/p1663354445071699 my understanding was that parsing any of the controller deployments would be anti-pattern here.

Reply:
Hmm, yeah I agree rukpak is not really an operand (more of a sibling). Having read the CVO docs you linked. I guess the core PO ClusterOperator doesn't really have an operand -- we've explicitly set the operands apart in the aggregated CO.

> An operator reports Progressing when it is rolling out new code, propagating config changes, or otherwise moving from one steady state to another. It should not report progressing when it is reconciling a previously known state. If it is progressing to a new version, it should include the version in the message for the condition like "Moving to v1.0.1".

It sounds like we'd set `Progressing=true` in the core CO when were performing PO upgrades during a cluster upgrade.

> An operator reports a new version when it has rolled out the new version to all of its operands.

And that also sounds like the point at which we'd set back to `Progressing=false`

Reply:
Do you think it's worth tinkering with P=[T,F] toggling during this phase given we implicitly don't support upgrades given the tech preview no upgrade feature set? Maybe that feature set only blocks minor version upgrades vs. blocks patch version upgrades.

A trivial implementation may compare the last recorded and current release version information that could be read from the Pod's `RELEASE_IMAGE` environment variable, and compare whether there's any skew.

Reply:
Talked about this in slack and the decision was to add a TODO comment for addressing this in future phases.